block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -53,49 +54,112 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
       (blockHash.size == BlockHash.Length && dagSet.contains(blockHash)).pure[F]
 
     def children(blockHash: BlockHash): F[Option[Set[BlockHash]]] =
-      childMap.get(blockHash).pure[F]
+      connectionsMap.get(blockHash).map(_.children intersect dagSet).pure[F]
 
     def latestMessageHash(validator: Validator): F[Option[BlockHash]] =
       latestMessagesMap.get(validator).pure[F]
 
-    def latestMessageHashes: F[Map[Validator, BlockHash]] = latestMessagesMap.pure[F]
+    def latestMessageHashes: F[Map[Validator, BlockHash]] =
+      latestMessagesMap.pure[F]
 
-    def invalidBlocks: F[Set[BlockMetadata]] = invalidBlocksSet.pure[F]
+    def invalidBlocks: F[Set[BlockMetadata]] =
+      invalidBlocksSet.filter(m => dagSet.contains(m.blockHash)).pure[F]
 
     // latestBlockNumber, topoSort and lookupByDeployId are only used in BlockAPI.
     // Do they need to be part of the DAG current state or they can be moved to DAG storage directly?
 
-    private def getMaxHeight = if (heightMap.nonEmpty) heightMap.last._1 + 1L else 0L
+    private def getMaxHeight: F[Long] = (if (heightMap.nonEmpty) heightMap.last._1 else 0L).pure[F]
 
-    def latestBlockNumber: F[Long] =
-      getMaxHeight.pure[F]
+    def latestBlockNumber: F[Long] = getMaxHeight
 
     def isFinalized(blockHash: BlockHash): F[Boolean] =
       finalizedBlocksSet.contains(blockHash).pure[F]
 
     def topoSort(
         startBlockNumber: Long,
         maybeEndBlockNumber: Option[Long]
-    ): F[Vector[Vector[BlockHash]]] = {
-      val maxNumber   = getMaxHeight
-      val startNumber = Math.max(0, startBlockNumber)
-      val endNumber   = maybeEndBlockNumber.map(Math.min(maxNumber, _)).getOrElse(maxNumber)
-      if (startNumber >= 0 && startNumber <= endNumber) {
-        Sync[F].delay(
-          heightMap
-            .filterKeys(h => h >= startNumber && h <= endNumber)
-            .map { case (_, v) => v.toVector }
-            .toVector
-        )
-      } else {
-        Sync[F].raiseError(
-          TopoSortFragmentParameterError(startNumber, endNumber)
-        )
+    ): F[Vector[Vector[BlockHash]]] =
+      getMaxHeight.flatMap { maxNumber =>
+        val startNumber = Math.max(0, startBlockNumber)
+        val endNumber   = maybeEndBlockNumber.map(Math.min(maxNumber, _)).getOrElse(maxNumber)
+        if (startNumber >= 0 && startNumber <= endNumber) {
+          Sync[F].delay(
+            heightMap
+              .filterKeys(h => h >= startNumber && h <= endNumber)
+              .map { case (_, v) => v.toVector.filter(dagSet.contains) }
```

Comment:
Filtering of hashes in height map can be done as part of truncate.

Reply:
We need to filter the whole height map, e.g. size of a million, on each DAG representation calculation. And here we only need to filter small part of this map which is of interest. Do you think it worth it to filter the whole map?

Reply:
We have block heights of latest messages so this can be used to restrict going to far in the past.

Reply:
I implemented truncate by only heights, and still not fully convinced - are you sure that if message A has height lower then lowest latest message - this A is always in view?